vocabularies/Common.json

```diff
@@ -1288,6 +1288,132 @@
       "@Common.Experimental": true,
       "@Core.IsURL": true,
       "@Core.Description": "Base URL for WebSocket connections"
+    },
+    "RecommendationsFunction": {
+      "$Kind": "Term",
+      "$Type": "Common.QualifiedName",
+      "$AppliesTo": ["EntityType"],
+      "@Common.Experimental": true,
+      "@Core.Description": "Qualified name of a function that computes AI-based recommendations for entities of the annotated type (see [template](#Template_RecommendationsFunction) and [template for asynchronous OData requests](#Template_AsyncRecommendationsFunction))"
+    },
+    "Template_RecommendationsFunction": [
+      {
+        "$Kind": "Function",
+        "$IsBound": true,
+        "@Common.Experimental": true,
+        "@Core.Description": "Template for functions that compute AI-based recommendations for the bound entity and its related entities",
+        "@Core.LongDescription": "If the function is invoked repeatedly in a stateful session, it SHOULD NOT\nrecompute recommendations whose predictors have not changed since the previous invocation.\n\nThe template function itself cannot be invoked.",
+        "$Parameter": [
+          { "$Name": "Entity", "$Type": "Edm.EntityType" },
+          {
+            "$Name": "Targets",
+            "$Collection": true,
+            "@Core.IsURL": true,
+            "@Core.Description": "URLs (relative to the request URL) that address entities, properties or action/function parameters for which recommendations shall be retrieved",
+            "@Core.LongDescription": "The request `SalesOrder('A')/ns.RecommendationsFunction?Targets=[\"Items(10)\"]`\n            retrieves recommendations for the entity `SalesOrder('A')/Items(10)`."
```

Comment:
If recommendations are to be fetched to Sales Order A itself then I assume we will make a request like below? or may be have no targets at all.

SalesOrder('A')/ns.RecommendationsFunction?Targets=[\"SalesOrder(A)\"]

and for navigation properties on Item the calls will look like below.

SalesOrder('A')/ns.RecommendationsFunction?Targets=[\"_Item(10)/_ItemPartner(20)\"]

Reply:
> If recommendations are to be fetched to Sales Order A itself then I assume we will make a request like below? or may be have no targets at all.

Either no `Targets` or `Targets=["."]`.

Reply:
An empty string, i.e. `Targets=[""]`, for targeting the binding parameter?

Reply:
You are right, the empty string is correct.
```
> new URL("", "http://server/service/SalesOrder('A')").href
"http://server/service/SalesOrder('A')"
> new URL(".", "http://server/service/SalesOrder('A')").href
'http://server/service/'
```

Reply:
From a modeling perspective does it really make sense to have SalesOrder('A')/ns.RecommendationsFunction?Targets=["_Item(10)/_ItemPartner(20)"] when you could just have SalesOrder('A')/_Item(10)/_ItemPartner(20)/ns.RecommendationsFunction ?

If you only consider the bindingParameter case does it really make sense to even specify the targets ? The AI Model generally generate recommendation for a full Tuple regardless of what you would specify (or at least they would have an opinion), as such i'm not even sure Targets are really relevant, or at least i miss the use case

Reply:
The targets allow you to bundle the calls for the main object and several items into one function call instead of calling the bound function for each individually. That may help the backend to have a better performance.

Reply:
And bundling is essential for the asynchronous behavior we want to achieve.

Reply:
I would be very careful with asynchronous request for something like this, especially if you consider bundling is important.
Let me take an example as to what we will be faced with now

User make change on MainObject and realize we need to fetch recommendation for multiple target

First we have

GET MainObject/ns.GetReco(Targets=["","_Item(10)","_Item(20)"]) --> Handle1

This gives us back Handle1 to fetch the recommendation at some point but this is long running

Meanwhile the frontend makes another change and end up doing a call with less or different targets

GET MainObject/ns.GetReco(Targets=["", "_Item(30)]) --> Handle2


Now you can't really delete Handle1, you still need to fetch it because it contains data that Handle2 won't but if by some chance Handle2 comes back before Handle1 (and this will happen because it can), then what do we do ? 

Do we overwrite the result when Handle1 come back ? Do we need to keep track of all request and have the client needing to determine per target what is relevant ? 

The whole asynchronous operations makes everything more complicated and I fail to see any benefit highlighted (beside the singular sticky case which I imagine has other possible solution)

Reply:
The client knows that the recommendations obtained via Handle1 must be processed before the ones obtained via Handle2. For now recommendations are too slow to be processed synchronously.